### PR TITLE
log-backup: let the advancer listen on the same range with the task

### LINKED ISSUE #37822: PiTR task cannot get checkpoint ts
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. start a  PiTR task

### 2. What did you expect to see? (Required)
1.  checkpoint ts can be advanced
2. `br log status` can get normal result
### 3. What did you see instead (Required)
1. checkpoint ts was not advanced
image
2. `br log status` got abnormal result
image
### 4. What is your TiDB version? (Required)

* br: 
```
/ # /br --version
Release Version: v6.2.0-alpha
Git Commit Hash: cdcbfc838ab3d4768d5761181ca9f3ee361811ce
Git Branch: heads/refs/tags/v6.2.0-alpha
Go Version: go1.18.5
UTC Build Time: 2022-09-08 11:01:32
Race Enabled: false
```
* tikv:
```
/ # /tikv-server --version
TiKV
Release Version:   6.2.0-alpha
Edition:           Community
Git Commit Hash:   8c93b91fea7396b9261551297d1bfa037c82e605
Git Commit Branch: heads/refs/tags/v6.2.0-alpha
UTC Build Time:    2022-09-13 11:02:56
Rust Version:      rustc 1.64.0-nightly (0f4bcadb4 2022-07-30)
Enable Features:   pprof-fp jemalloc mem-profiling portable sse test-engine-kv-rocksdb test-engine-raft-raft-engine cloud-aws cloud-gcp cloud-azure
Profile:           dist_release
```

### PR DESCRIPTION
**Pull request by YuJuncen**
### What problem does this PR solve?

Issue Number: close #37822

Problem Summary:
Perviously, the advancer would listen the full key space and wait all regions to get flushed then advancing the checkpoint.
In some condition (Maybe related to DR auto sync), PD may split at the key `m`, which leads to an always empty region ["", "m"). This region would never be observed by log-backup hence would never get flushed. Which leads to the checkpoint doesn't get advanced.

### What is changed and how it works?
This (huge) PR allows the advancer to listen the same range with the task.
Generally, it added a new field `taskRange` to the `Advancer`, and:
- For `CalculateGlobalCheckpoint`, it uses the task range as the initial range.
- For `CalculateGlobalCheckpointLight`, it would "clamp" the collected ranges within the task ranges. See the `ClampRanges` for details.
- For `ConsistencyCheck`, it would check the current ranges with the task range.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fixed a bug that may cause log backup checkpoint TS stuck when some weird ranged regions exist. 
```

### CODE REVIEW

File: br/pkg/task/test_main/main.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package main
```
Comment (3pointer) [inline br/pkg/task/test_main/main.go:1]: add license

Comment (YuJuncen) [reply br/pkg/task/test_main/main.go:1]: Oops, this file should be removed.

File: br/pkg/utils/key.go [hunk 2]
```diff
@@ -89,3 +94,92 @@ func CompareEndKey(a, b []byte) int {
...
+	leftNotOverlapped
+	// ToClamp :    |_________|
+	// Range:                    |______|
+	rightNotOverlapped
+	buggyUnknown
+)
+
+func clampInOneRange(rng kv.KeyRange, clampIn kv.KeyRange) (kv.KeyRange, failedToClampReason) {
+	possibleFailureReason := buggyUnknown
+	if bytes.Compare(rng.StartKey, clampIn.StartKey) < 0 {
```
Comment (Leavrth) [inline br/pkg/utils/key.go:131]: startkeys use bare compare here while tsheap.go:285 use the function `CompareBytesExt`. Maybe unify them have better readability.

**Review by Leavrth — APPROVED** [review]

File: br/pkg/streamhelper/advancer.go [hunk 3]
```diff
@@ -193,12 +194,15 @@ func (c *CheckpointAdvancer) tryAdvance(ctx context.Context, rst RangesSharesTS)
...
+		return rst.Ranges[i]
+	})
 	workers := utils.NewWorkerPool(4, "sub ranges")
 	eg, cx := errgroup.WithContext(ctx)
 	collector := NewClusterCollector(ctx, c.env)
 	collector.setOnSuccessHook(c.cache.InsertRange)
-	for _, r := range ranges {
+	clampedRanges := utils.ClampRanges(ranges, c.taskRange)
+	for _, r := range clampedRanges {
 		r := r
```
Comment (Leavrth) [inline br/pkg/streamhelper/advancer.go:206]: why r := r 🤔

Comment (YuJuncen) [reply br/pkg/streamhelper/advancer.go:206]: Just let the closure capture the right value...

**Review by 3pointer — APPROVED** [review]

**Pull request merged**